### PR TITLE
Participant / Sample External IDs in parser

### PR DESCRIPTION
**Pull request by illusional**
Add support for multiple participant and external IDs in the parsers, in two ways:

- Allow some subclass to provide an implementation for `get_{participant/sample}_external_ids`.
- Add an implementation here for the `GenericMetadataParser`.
- Switching matching participants / samples to look through ALL IDs.
- Add a docstring to better describe parameters.
- Fix a bug in update parent / root for nested samples.
-

### CODE REVIEW

File: db/python/layers/sample.py [hunk 1]
```diff
@@ -243,8 +243,8 @@ async def upsert_sample(
                         participant_id=s.participant_id,
                         type_=s.type,
                         active=s.active,
-                        sample_parent_id=sample_parent_id,
-                        sample_root_id=sample_root_id,
+                        sample_parent_id=r.parent.id if r.parent else sample_parent_id,
```
Comment (illusional) [inline db/python/layers/sample.py:246]: Not strictly relevant to this PR, but a bug from #840

File: metamist/parser/generic_parser.py [hunk 2]
```diff
@@ -963,33 +988,52 @@ async def validate_rows(self, rows: list[SingleRow]):
...
+    def get_reported_gender(self, row: GroupedRow) -> str | None:
         """Get reported gender from grouped row"""
         return None
 
-    def get_karyotype(self, row: GroupedRow) -> Optional[str]:
+    def get_karyotype(self, row: GroupedRow) -> str | None:
         """Get karyotype from grouped row"""
         return None
 
     # @abstractmethod
```
Comment (illusional) [inline metamist/parser/generic_parser.py:1033]:
```suggestion
```

File: metamist/parser/generic_parser.py [hunk 3]
```diff
@@ -190,6 +184,22 @@ def process_fieldname(self, fieldname: str) -> str:
         return fieldname.strip()
 
 
+def prepare_external_ids(
+    primary_external_id: str, external_ids: dict[str, str]
```
Comment (jmarshall) [inline metamist/parser/generic_parser.py:188]: If we need `or {}` below, then this probably wants to be `dict[str, str] | None`

File: metamist/parser/generic_parser.py [hunk 4]
```diff
@@ -198,15 +208,17 @@ def __init__(
...
+        reported_sex: int | None,
+        reported_gender: str | None,
+        karyotype: str | None,
+        external_pids: dict[str, str] | None = None,
     ):
         self.rows = rows
 
         self.internal_pid = internal_pid
         self.external_pid = external_pid
+        self.external_pids = external_pids or {}
```
Comment (jmarshall) [inline metamist/parser/generic_parser.py:221]: Might deserve a comment noting that — unlike other layers — the primary one isn't in `external_pids` in this class.

File: metamist/parser/generic_parser.py [hunk 5]
```diff
@@ -239,13 +251,16 @@ def __init__(
...
-        meta: Dict[str, Any] | None = None,
+        meta: dict[str, Any] | None = None,
+        external_sids: dict[str, str] | None = None,
     ):
         self.participant = participant
         self.rows = rows
 
         self.internal_sid = internal_sid
         self.external_sid = external_sid
+        self.external_sids = external_sids or {}
```
Comment (jmarshall) [inline metamist/parser/generic_parser.py:262]: Ditto re comment

File: metamist/parser/generic_parser.py [hunk 6]
```diff
@@ -667,9 +682,9 @@ async def from_json(self, rows, confirm=False, dry_run=False):
 
     def _get_dict_reader(self, file_pointer, delimiter: str):
         """
-        Return a DictReader from file_pointer
+        Return a dictReader from file_pointer
```
Comment (jmarshall) [inline metamist/parser/generic_parser.py:685]: stet here and on the line a couple below

File: db/python/layers/sample.py (hunk 1 above)
Comment (jmarshall) [reply db/python/layers/sample.py:246]: Fair enough!

**Review by jmarshall — APPROVED** [review]
Some minor comments, otherwise looks good to me.

Are the codecov notes complaining about missing test coverage on the new `get_participant_external_ids()` etc functions accurate?

**Pull request merged**